OpenTrons/opentrons-api
# refactor(app-shell): Migrate desktop app notify connectivity to discovery-client

Closes [EXEC-304](https://opentrons.atlassian.net/browse/EXEC-304)



# Overview

Migrating MQTT connectivity logic to discovery-client simplifies  app-shell connectionStore manager logic (maybe), reduces bug surface, and gives us connection retry logic relatively for free. 

Most importantly, we will not depend on components to express interest in a broker and topics, which implicitly requires all components and notify helper functions in the app to behave appropriately (not re-render like crazy). 

Note that there's some browser layer logic that is no longer relevant (like UNSUBSCRIBE actions), but these still need to be here until app-shell-odd is refactored. While the app still dispatches these actions, they don't actively do anything.

App-shell notifications behave as follows:
- Broker connections are added or forcefully removed based on health status changes reported by discovery-client.
- Subscriptions are handled "lazily", in which a component must express interest in a topic before a subscription request is made (no network request will be made if already subscribed tot he topic). 
- Unsubscribe requests only occur if an "unsubscribe" flag is received from the broker. That means the only thing we actively unsubscribe from at the moment is "/runs/:runId".
- Pending subs and unsubs are used to prevent unnecessary network and broker load.
- A blocked port event can only be reported a maximum of one time per robot per app session (no change here). 



# Test Plan
- Push https://github.com/Opentrons/opentrons/pull/14620 to a robot. 
- The easiest way to test this is to do a `setInterval(() => {console.log(connectionStore)}, 1000)` within `notify.ts` as this gives direct feedback on what's happening.
- Start the app in dev mode, you should see a bunch of app-shell logs in which the app tries to connect to all robots that are _available_. Note that on our network, there are a lot of OT-2s and Flexes on 7.0.2, so it's fine to see a bunch of failed connections. Some should still go through through.
- Click on the Devices page. In the app-shell logs, you should see a bunch of subscribe actions going through the IPC and the app-shell should report successful subscribes. 
- Turn off a robot. Whenever the discovery-client polls and updates the available robots tab, you should see that robot be removed from the connection store. You won't actually see the "closed connection" log, because the robot is completely offline.
- Run a protocol and cancel it. I don't remember if it's when you restart a run or as soon as the cancel process is complete, but at some point the app-shell logs should say "successfully unsubscribed..." with the topic string that matches the run ID (this is the only topic with an unsubscribe flag for now). 







# Risk assessment
medium-ish. This is a pretty big refactor of how notifications work for the desktop app. 



[EXEC-304]: https://opentrons.atlassian.net/browse/EXEC-304?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
